### PR TITLE
feat: Use timestamp of last block when dry running transactions

### LINKED ISSUE #2128: Manually produces blocks in test and simulates a call the timestamp is not correct
Maybe related to https://github.com/FuelLabs/fuel-core/issues/2091

https://fuellabs.slack.com/archives/C01L1HUJSL8/p1724242020509839

### PR DESCRIPTION
**Pull request by netrome**
## Description
Closes #2128 

This PR adds the option to specify a timestamp in BlockProducer::dry_run. When no timestamp is provided, the block producer will use the time stamp of the previous block, instead of using `Tai64::now()` (previous behavior).

## Checklist
- [x] Breaking changes are clearly marked as such in the PR description and changelog
- [x] New behavior is reflected in tests

### Before requesting review
- [x] I have reviewed the code myself
- [x] I have created follow-up issues caused by this PR and linked them here

### After merging, notify other teams

- [ ] [Rust SDK](https://github.com/FuelLabs/fuels-rs/)

### CODE REVIEW

File: crates/services/producer/src/block_producer/tests.rs [hunk 1]
```diff
@@ -543,6 +543,86 @@ mod produce_and_execute_block_txpool {
     }
 }
 
+// Tests for the `dry_run` method.
+mod dry_run {
+    use super::*;
+
+    #[tokio::test]
+    async fn dry_run__executes_with_given_timestamp() {
```
Comment (MitchTurner) [inline crates/services/producer/src/block_producer/tests.rs:551]: What if time is earlier than the last block? Is that valid?

File: crates/services/producer/src/mocks.rs [hunk 2]
```diff
@@ -227,6 +229,20 @@ impl BlockProducer<Vec<Transaction>> for MockExecutorWithCapture<Transaction> {
...
 
+impl DryRunner for MockExecutorWithCapture<Transaction> {
+    fn dry_run(
+        &self,
+        block: Components<Vec<Transaction>>,
+        _utxo_validation: Option<bool>,
+    ) -> ExecutorResult<Vec<TransactionExecutionStatus>> {
+        *self.captured.lock().unwrap() = Some(block);
+
+        Err(ExecutorError::Other(
```
Comment (xgreenx) [inline crates/services/producer/src/mocks.rs:240]: Why do we want to return an error? It seems will be more clear to retune `Ok`, since you verify the timestamp from the `block`

File: crates/services/producer/src/block_producer/tests.rs [hunk 3]
```diff
@@ -865,67 +956,49 @@ impl TestContextBuilder {
...
+        let block_gas_limit = self.block_gas_limit.unwrap_or_default();
 
         let mock_relayer = MockRelayer {
             latest_block_height: self.latest_block_height,
             latest_da_blocks_with_costs: self.blocks_with_gas_costs.clone(),
             ..Default::default()
         };
 
+        let db = MockDb {
+            blocks: self.pre_existing_blockss(),
```
Comment (xgreenx) [inline crates/services/producer/src/block_producer/tests.rs:994]:
```suggestion
            blocks: self.pre_existing_blocks(),
```

File: tests/tests/tx.rs [hunk 4]
```diff
@@ -252,6 +261,64 @@ async fn submit_and_await_status() {
...
+        .result
+    else {
+        panic!("unexpected execution result from dry run")
+    };
+
+    // Then
+    assert_eq!(
+        start_time.0
+            + block_production_interval_seconds
+                * number_of_blocks_to_produce_manually.saturating_sub(1) as u64,
```
Comment (xgreenx) [inline tests/tests/tx.rs:317]: Can we move it to a separate variable?

File: crates/services/producer/src/mocks.rs (hunk 2 above)
Comment (netrome) [reply crates/services/producer/src/mocks.rs:240]: My thinking here is that this mock **only provides a side channel to capture the block** right now, and it **does not simulate any execution**. Because of this, I prefer to return an error here to clarify for the caller that the simulated execution failed because it's not implemented. This should prevent any future misconceptions if someone uses this mock and assumes it actually simulates execution.

If you still prefer to have this return an `Ok` result, I'm also up for doing that. Let me know what you think.

File: crates/services/producer/src/block_producer/tests.rs (hunk 3 above)
Comment (netrome) [reply crates/services/producer/src/block_producer/tests.rs:994]: Hahah, sounds like a snake talking :sweat_smile: I'll fix. Thanks for spotting!

File: tests/tests/tx.rs (hunk 4 above)
Comment (netrome) [reply tests/tests/tx.rs:317]: Sure thing, will do.

File: crates/services/producer/src/block_producer/tests.rs (hunk 1 above)
Comment (netrome) [reply crates/services/producer/src/block_producer/tests.rs:551]: Good question, I'd assume that it's valid because it would allow for testing contrafactuals. I.e. "what if this transaction had been submitted in the past, would it have been valid?". Though we do error on previous block heights (which we shouldn't do though https://github.com/FuelLabs/fuel-core/issues/2062).

Comment (netrome) [reply crates/services/producer/src/block_producer/tests.rs:551]: I'll add a test case for it.

File: tests/tests/tx.rs (hunk 4 above)
Comment (netrome) [reply tests/tests/tx.rs]: Done in https://github.com/FuelLabs/fuel-core/pull/2206/commits/c8f3ee1b540bf35a0cf2fc3cd8ed9fa48bb9223e

File: crates/services/producer/src/block_producer/tests.rs (hunk 3 above)
Comment (netrome) [reply crates/services/producer/src/block_producer/tests.rs]: Fixed in https://github.com/FuelLabs/fuel-core/pull/2206/commits/c8f3ee1b540bf35a0cf2fc3cd8ed9fa48bb9223e

File: crates/services/producer/src/block_producer/tests.rs (hunk 1 above)
Comment (netrome) [reply crates/services/producer/src/block_producer/tests.rs:551]: Added in https://github.com/FuelLabs/fuel-core/pull/2206/commits/c8f3ee1b540bf35a0cf2fc3cd8ed9fa48bb9223e and it works.

File: crates/services/producer/src/mocks.rs (hunk 2 above)
Comment (MitchTurner) [reply crates/services/producer/src/mocks.rs:240]: I think the `Err` is noise. It's harder to understand the tests since an error feels like it should be an edge case.

Admittedly, any "happy" value will also feel arbitrary, but since you aren't inspecting them it won't feel like extra context.

something like `arb_succeeding` instead of `default` might help communicate, but is probably overkill.

Comment (netrome) [reply crates/services/producer/src/mocks.rs:240]: Alright, fair enough. I'll change to return an `Ok` result then.

Comment (netrome) [reply crates/services/producer/src/mocks.rs]: Done in https://github.com/FuelLabs/fuel-core/pull/2206/commits/c3661969e331692399d064f2d446b2c4bd594dc3

File: crates/services/producer/src/block_producer.rs [hunk 5]
```diff
@@ -268,26 +268,33 @@ where
...
+
+        let simulated_height = height.unwrap_or_else(|| {
+            latest_height
                 .succ()
                 .expect("It is impossible to overflow the current block height")
         });
 
-        let header = self._new_header(height, Tai64::now())?;
+        let simulated_time = time.unwrap_or_else(|| {
+            view.get_block(&latest_height)
```
Comment (rymnc) [inline crates/services/producer/src/block_producer.rs:292]: qq: should this use the latest height or the simulated height?

Comment (netrome) [reply crates/services/producer/src/block_producer.rs:292]: Simulated height might not have a block, so defaulting to latest makes sense here to me. If the caller specifies a block height, and not a block time it is likely that the thing they are testing isn't using the block time. If someone needs to simulate a block with a specific height and timestamp, they can provide both.

File: crates/services/producer/src/block_producer/tests.rs [hunk 6]
```diff
@@ -543,6 +543,110 @@ mod produce_and_execute_block_txpool {
...
+        let simulated_block_time = Tai64::from_unix(1337);
+        let executor = MockExecutorWithCapture::default();
+        let ctx = TestContext::default_from_executor(executor.clone());
+
+        // When
+        let _err = ctx
+            .producer()
+            .dry_run(vec![], None, Some(simulated_block_time), None, None)
+            .await
+            .expect("dry run failed");
```
[the remark this answers is not in the document]
Comment (netrome) [reply crates/services/producer/src/block_producer/tests.rs:562]: Since I changed the executor to return `Ok`, this is a normal `expect` so it should only panic if the dry run has failed. Though thinking about this we don't care about whether this returns an error or not here so I'll change this to just ignore the results instead.

[the remark this answers is not in the document]
Comment (netrome) [reply crates/services/producer/src/block_producer/tests.rs]: Changed in https://github.com/FuelLabs/fuel-core/pull/2206/commits/24d509552fadf33412bb3bc488ebac738c839f96

File: crates/services/producer/src/block_producer/tests.rs [hunk 7]
```diff
@@ -543,6 +543,110 @@ mod produce_and_execute_block_txpool {
...
+        let ctx = TestContextBuilder::new()
+            .with_prev_time(last_block_time)
+            .build_with_executor(executor.clone());
+
+        // When
+        let _err = ctx
+            .producer()
+            .dry_run(vec![], None, None, None, None)
+            .await
+            .expect("dry run failed");
```
[the remark this answers is not in the document]
Comment (netrome) [reply crates/services/producer/src/block_producer/tests.rs]: Changed to ignore the result in https://github.com/FuelLabs/fuel-core/pull/2206/commits/24d509552fadf33412bb3bc488ebac738c839f96

File: crates/services/producer/src/block_producer/tests.rs [hunk 8]
```diff
@@ -543,6 +543,110 @@ mod produce_and_execute_block_txpool {
...
+        let ctx = TestContextBuilder::new()
+            .with_prev_time(last_block_time)
+            .build_with_executor(executor.clone());
+
+        // When
+        let _err = ctx
+            .producer()
+            .dry_run(vec![], None, Some(simulated_block_time), None, None)
+            .await
+            .expect("dry run failed");
```
[the remark this answers is not in the document]
Comment (netrome) [reply crates/services/producer/src/block_producer/tests.rs]: Changed to ignore the result in https://github.com/FuelLabs/fuel-core/pull/2206/commits/24d509552fadf33412bb3bc488ebac738c839f96

**Review by xgreenx — DISMISSED** [review]

**Review by xgreenx — DISMISSED** [review]

**Review by rymnc — DISMISSED** [review]

**Review by rafal-ch — APPROVED** [review]
LGTM :+1:

File: crates/services/producer/src/block_producer.rs [hunk 9]
```diff
@@ -268,26 +268,33 @@ where
...
+                .map(|block| block.header().time())
+                .unwrap_or(Tai64::UNIX_EPOCH)
+        });
+
+        let header = self._new_header(simulated_height, simulated_time)?;
```
[the remark this answers is not in the document]
Comment (netrome) [reply crates/services/producer/src/block_producer.rs:297]: Yeah I'm wondering the same thing :sweat_smile:

[the remark this answers is not in the document]
Comment (netrome) [reply crates/services/producer/src/block_producer.rs:297]: Seems to be to avoid a collision with the existing `new_header` function but yeah those methods could be named better.

[the remark this answers is not in the document]
Comment (netrome) [reply crates/services/producer/src/block_producer.rs]: Renamed to better reflect what the functions seem to be doing in https://github.com/FuelLabs/fuel-core/pull/2206/commits/98c32af83bedc1135ec204ab25235241108c44b4
```
new_header -> new_header_with_new_da_height
_new_header -> new_header
```

**Review by rafal-ch — APPROVED** [review]

**Review by MitchTurner — APPROVED** [review]
